### PR TITLE
[Parquet | ExportDatabase] Deal with unsupported parquet types in EXPORT DATABASE.

### PR DESCRIPTION
**Pull request by Tishj**
### Write as VARCHAR

We don't support writing some types as Parquet (yet).
To make sure this does not blow up in export/import database, we fall back to VARCHAR for types that are not supported.
Relying on the ability to roundtrip the values to and from string.

### HUGEINT
~~Hugeint was previously written as DOUBLE, but when testing with `test_all_types()` this results in a conversion exception on IMPORT, so I've also removed support there and fall back to VARCHAR there as well.~~
Hugeint is a supported type by the parquet writer, but it's lossy, so for EXPORT DATABASE we will fall back to VARCHAR.

### UNION

Since UNION loses information on string cast, we add additional logic for them.
When exporting a database to parquet containing a UNION column, it is now written as STRUCT instead, including the tag information

Since EXPORT DATABASE writes a `schema.sql` we don't lose the type information.
We allow casting a struct to UNION if the struct value has a valid representation for UNION internals

### CODE REVIEW

**Comment by Tishj** [comment]
As pointed out by Carlo, this does have a flaw when typed nulls are involved:

```sql
statement ok
begin transaction;

statement ok
create table tbl2 (a UNION(a bit, b bool));

statement ok
insert into tbl2 VALUES
	(NULL),
	(union_value(a := NULL)),
	(union_value(b := NULL));

statement ok
SELECT union_tag(a) FROM tbl2;
#┌────────────────┐
#│  union_tag(a)  │
#│ enum('a', 'b') │
#├────────────────┤
#│                │
#│ a              │
#│ b              │
#└────────────────┘

statement ok
EXPORT DATABASE '__TEST_DIR__/union' (FORMAT PARQUET);

statement ok
rollback;

statement ok
IMPORT DATABASE '__TEST_DIR__/union';

statement ok
SELECT union_tag(a) FROM tbl2;
#┌────────────────┐
#│  union_tag(a)  │
#│ enum('a', 'b') │
#├────────────────┤
#│                │
#│                │
#│                │
#└────────────────┘
```

File: extension/parquet/parquet_writer.cpp [hunk 1]
```diff
@@ -94,8 +95,11 @@ bool ParquetWriter::DuckDBTypeToParquetTypeInternal(const LogicalType &duckdb_ty
 		parquet_type = Type::FLOAT;
 		break;
 	case LogicalTypeId::DOUBLE:
+		parquet_type = Type::DOUBLE;
+		break;
 	case LogicalTypeId::HUGEINT:
 		parquet_type = Type::DOUBLE;
+		return CopyTypeSupport::LOSSY;
 		break;
```
Comment (Mytherin) [inline extension/parquet/parquet_writer.cpp:103]: nit: the break here can go

File: src/common/types.cpp [hunk 2]
```diff
@@ -400,7 +400,7 @@ string LogicalType::ToString() const {
 			if (i > 0) {
 				ret += ", ";
 			}
-			ret += "'" + KeywordHelper::WriteOptionallyQuoted(EnumType::GetString(*this, i).GetString(), '\'') + "'";
```
Comment (Mytherin) [inline src/common/types.cpp:403]: I think this breaks round-tripping of ENUM types - is this change necessary in this PR?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good - two minor comments.

Could you also have a look into fully covering `from_struct.cpp` in the tests?

File: src/common/types.cpp (hunk 2 above)
Comment (Tishj) [reply src/common/types.cpp:403]:
```
Query unexpectedly failed! (test/sql/export/parquet/export_parquet_enum.test:33)!
================================================================================
IMPORT DATABASE 'duckdb_unittest_tempdir/18323/export_test';
================================================================================
Parser Error: syntax error at or near "1"
LINE 5: CREATE TABLE tbl(a ENUM(''1'', ''2'', ''3''));
```

When I run the `export_parquet_enum.test` on `main`

Comment (Tishj) [reply src/common/types.cpp:403]: Since we already directly add singlequotes around it, I don't think it's necessary to add WriteOptionallyQuoted here, right?
@Mytherin

Comment (Mytherin) [reply src/common/types.cpp:403]: Oh yeah you're right, I missed the explicit adding of quotes. I think we still need to deal with strings that contain `'` explicitly though. Could you add a test with an enum like this (`select 'hello''world'::enum('hello''world');`) - I think we should use `KeywordHelper::WriteQuotes` instead of adding the quotes

**Comment by Tishj** [comment]
> Thanks for the PR! Looks good - two minor comments.
> 
> Could you also have a look into fully covering `from_struct.cpp` in the tests?

Some of them are already asserted by the `AllowImplicitCastFromStruct` function
such as:
```c++
		if (!cast_data.child_cast_info[i].function(source_child_vector, result_child_vector, count, child_parameters)) {
			all_converted = false;
		}
```
We have already made sure the types are the same

```c++
		if (entry.init_local_state) {
			CastLocalStateParameters child_params(parameters, entry.cast_data);
			child_state = entry.init_local_state(child_params);
		}
```
I haven't seen anything use this when a direct cast is made, but I'll have another look

**Comment by Mytherin** [comment]
I think the JSON casts use the local state

**Comment by Mytherin** [comment]
Thanks!
